Libraries/LibKeyboard/CharacterMapFile.cpp

```diff
@@ -0,0 +1,106 @@
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "CharacterMapFile.h"
+#include <LibCore/File.h>
+
+namespace LibKeyboard {
+
+Optional<CharacterMapData> CharacterMapFile::load_from_file(const String& file_name)
+{
+    auto path = file_name;
+    if (!path.ends_with(".json")) {
+        StringBuilder full_path;
+        full_path.append("/res/keymaps/");
+        full_path.append(file_name);
+        full_path.append(".json");
+        path = full_path.to_string();
+    }
+
+    auto file = Core::File::construct(path);
+    file->open(Core::IODevice::ReadOnly);
+    if (!file->is_open()) {
+        dbg() << "Failed to open " << file_name << ":" << file->error_string();
+        return {};
+    }
+
+    auto file_contents = file->read_all();
+    auto json = JsonValue::from_string(file_contents).as_object();
+
+    char* map = read_map(json, "map");
+    char* shift_map = read_map(json, "shift_map");
+    char* alt_map = read_map(json, "alt_map");
+    char* altgr_map = read_map(json, "altgr_map");
+
+    CharacterMapData character_map;
+    for (int i = 0; i < CHAR_MAP_SIZE; i++) {
+        character_map.map[i] = map[i];
+        character_map.shift_map[i] = shift_map[i];
+        character_map.alt_map[i] = alt_map[i];
+        if (altgr_map) {
+            character_map.altgr_map[i] = altgr_map[i];
+        } else {
+            // AltGr map was not found, using Alt map as fallback.
+            character_map.altgr_map[i] = alt_map[i];
+        }
+    }
+
+    return character_map;
+}
+
+char* CharacterMapFile::read_map(const JsonObject& json, const String& name)
+{
+    if (!json.has(name))
+        return nullptr;
+
+    char* map = new char[CHAR_MAP_SIZE]();
+    auto map_arr = json.get(name).as_array();
+
+    for (int i = 0; i < map_arr.size(); i++) {
+        auto key_value = map_arr.at(i).as_string();
+        char character = 0;
+        if (key_value.length() == 0) {
+            ;
```

Comment:
:|

Reply:
I am not happy with this. JSON parser needs more interest for fix this.